packages/autoloader/src/AutoloadGenerator.php

```diff
@@ -0,0 +1,348 @@
+	 * @param PackageInterface             $mainPackage Main Package object.
+	 * @param InstallationManager          $installationManager Manager for installing packages.
+	 * @param string                       $targetDir Path to the current target directory.
+	 * @param bool                         $scanPsr0Packages Whether to search for packages. Currently hard coded to always be false.
+	 * @param string                       $suffix The autoloader suffix ignored since we want our autolaoder to only be included once.
+	 */
+	public function dump(
+		Config $config,
+		InstalledRepositoryInterface $localRepo,
+		PackageInterface $mainPackage,
+		InstallationManager $installationManager,
+		$targetDir,
+		$scanPsr0Packages = null, // Not used we always optimize.
+		$suffix = null // Not used since we create our own autoloader.
+	) {
+
+		$filesystem = new Filesystem();
+		$filesystem->ensureDirectoryExists( $config->get( 'vendor-dir' ) );
+
+		$basePath   = $filesystem->normalizePath( realpath( getcwd() ) );
+		$vendorPath = $filesystem->normalizePath( realpath( $config->get( 'vendor-dir' ) ) );
+		$targetDir  = $vendorPath . '/' . $targetDir;
+		$filesystem->ensureDirectoryExists( $targetDir );
+
+		$vendorPathCode = $filesystem->findShortestPathCode( realpath( $targetDir ), $vendorPath, true );
+
+		$appBaseDirCode = $filesystem->findShortestPathCode( $vendorPath, $basePath, true );
+		$appBaseDirCode = str_replace( '__DIR__', '$vendorDir', $appBaseDirCode );
+
+		$packageMap = $this->buildPackageMap( $installationManager, $mainPackage, $localRepo->getCanonicalPackages() );
+		$autoloads  = $this->parseAutoloads( $packageMap, $mainPackage );
+
+		$classMap = $this->getClassMapFromAutoloads( $autoloads, $filesystem, $vendorPath, $basePath );
+
+		// Write out the autoload_classmap_package.php file.
+		$classmapFile  = <<<EOF
+<?php
+
+// This file `autoload_packages.php`was generated by automattic/jetpack-autoloader.
+
+\$vendorDir = $vendorPathCode;
+\$baseDir = $appBaseDirCode;
+
+
+EOF;
+		$classmapFile .= 'return ' . $this->classMapToPHPArrayString( $classMap );
+		file_put_contents( $targetDir . '/autoload_classmap_package.php', $classmapFile );
+		$this->io->writeError( '<info>Generated autoload_classmap_package.php</info>', true );
+
+		if ( ! $suffix ) {
+			$suffix = $config->get( 'autoloader-suffix' )
+				? $config->get( 'autoloader-suffix' )
+				: md5( uniqid( '', true ) );
+		}
+		// Copy over the autoload.php file into autoload_packages.php.
+		file_put_contents( $vendorPath . '/autoload_packages.php', $this->getAutoloadPackageFile( $suffix ) );
+		$this->io->writeError( '<info>Generated autoload_packages.php</info>', true );
+
+	}
+
+	/**
+	 * Takes a classMap and returns the array string representation.
+	 *
+	 * @param array $classMap Map of all the package classes and paths and versions.
+	 *
+	 * @return string
+	 */
+	private function classMapToPHPArrayString( $classMap ) {
```

Comment:
Should we add a type hint here, to follow the rest of the class code style?

```suggestion
	private function classMapToPHPArrayString( array $classMap ) {
```